orca/producer.go

```diff
@@ -0,0 +1,222 @@
+	var once sync.Once
+	return func() {
+		once.Do(func() {
+			p.unregisterListener(l, opts.ReportInterval)
+			close()
+		})
+	}
+}
+
+type producer struct {
+	client v3orcaservicegrpc.OpenRcaServiceClient
+
+	closed *grpcsync.Event // fired when closure completes
+	// backoff is called between stream attempts to determine how long to delay
+	// to avoid overloading a server experiencing problems.  The attempt count
+	// is incremented when stream errors occur and is reset when the stream
+	// reports a result.
+	backoff func(int) time.Duration
+
+	mu        sync.Mutex
+	intervals map[time.Duration]int    // map from interval time to count of listeners requesting that time
+	listeners map[OOBListener]struct{} // set of registered listeners
+}
+
+// registerListener adds the listener and its requested report interval to the
+// producer.
+func (p *producer) registerListener(l OOBListener, interval time.Duration) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.listeners[l] = struct{}{}
+	p.intervals[interval]++
+}
+
+// registerListener removes the listener and its requested report interval to
+// the producer.
+func (p *producer) unregisterListener(l OOBListener, interval time.Duration) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	delete(p.listeners, l)
+	p.intervals[interval]--
+	if p.intervals[interval] == 0 {
+		delete(p.intervals, interval)
+	}
+}
+
+// minInterval returns the smallest key in p.intervals.
+func (p *producer) minInterval() time.Duration {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	var min time.Duration
+	for i, t := range p.intervals {
+		if t < min || i == 0 {
+			min = t
+		}
+	}
+	return min
+}
+
+// run manages the ORCA OOB stream on the subchannel.
+func (p *producer) run(ctx context.Context) {
+	defer p.closed.Fire()
+	backoffAttempt := 0
+	backoffTimer := time.NewTimer(0)
+	for ctx.Err() == nil {
+		select {
+		case <-backoffTimer.C:
+		case <-ctx.Done():
+			return
+		}
+
+		resetBackoff, err := p.runStream(ctx)
+
+		if resetBackoff {
+			backoffTimer.Reset(0)
+			backoffAttempt = 0
+		} else {
+			backoffTimer.Reset(p.backoff(backoffAttempt))
+			backoffAttempt++
+		}
+
+		switch {
+		case err == nil:
+			// No error was encountered; restart the stream.
+		case ctx.Err() != nil:
+			// Producer was stopped; exit immediately and without logging an
+			// error.
+			return
+		case status.Code(err) == codes.Unimplemented:
+			// Unimplemented; do not retry.
+			logger.Error("Server doesn't support ORCA OOB load reporting protocol; not listening for load reports.")
+			return
+		case status.Code(err) == codes.Unavailable:
+			// The SubConn is not currently ready; backoff silently.
+			//
+			// TODO: don't attempt the stream until the state is READY to
+			// minimize the chances of this case and to avoid using the
+			// exponential backoff mechanism, as we should know it's safe to
+			// retry when the state is READY again.
+		default:
+			// Log all other errors.
+			logger.Error("Received unexpected stream error:", err, ctx.Err())
```

Comment:
This is related to my other comment about resetBackoff return. io.EOF? Isn't the only way to return out of runStream() without a nil error if interval != p.minInterval()? Otherwise would always return and cause log spam on every completed stream...? Also this formatting is off. Add a space in between error: and err and add something explaining ctx.Err()? Right now there are two errors being logged. Is ctx even guaranteed to have an error in this case?